File: nextjs/int-reztro-restaurant-mgt-dashboard-web-dev/services/messageService.js
import api from "./api";

// Helpers to extract IDs dynamically safely
const getConversationId = (idOrObj) => {
    return idOrObj?.conversation_id || idOrObj?.id || idOrObj;
};

const getMessageId = (idOrObj) => {
    return idOrObj?.message_id || idOrObj?.id || idOrObj;
};

// Payload Builders for clean structures matching Swagger & SQL Schema
const buildConversationPayload = (payload) => {
    return {
        conversation_no: payload.conversation_no || null,
        participant_name: payload.participant_name?.trim() || "",
        participant_role: payload.participant_role?.trim() || "customer",
        customer_id: payload.customer_id ? Number(payload.customer_id) : null,
        admin_id: payload.admin_id ? Number(payload.admin_id) : null,
        avatar_url: payload.avatar_url || null,
        avatar_text: payload.avatar_text || null,
        is_online: payload.hasOwnProperty("is_online") ? payload.is_online : false,
        is_active: payload.hasOwnProperty("is_active") ? payload.is_active : true,
        last_message: payload.last_message || null,
        last_message_time: payload.last_message_time || null,
        unread_count: Number(payload.unread_count || 0),
        is_read: payload.hasOwnProperty("is_read") ? payload.is_read : false,
    };
};

const buildMessagePayload = (payload) => {
    return {
        conversation_id: Number(getConversationId(payload.conversation_id)),
        sender_id: String(payload.sender_id),
        sender_type: payload.sender_type || "admin", // admin, customer, chef, kitchen_admin etc.
        message_type: payload.message_type || "text",
        message_text: payload.message_text?.trim() || "",
        attachment_url: payload.attachment_url || null,
        attachment_type: payload.attachment_type || null,
        is_read: payload.hasOwnProperty("is_read") ? payload.is_read : false,
        is_active: payload.hasOwnProperty("is_active") ? payload.is_active : true,
    };
};

export const messageService = {
    /*
    |--------------------------------------------------------------------------
    | MESSAGE CONVERSATIONS CRUD (Sidebar & User Status)
    |--------------------------------------------------------------------------
    */

    // Get all conversations list (For sidebar display)
    getConversations: () => api.get("/message-conversations"),

    // Get single conversation profile details by ID
    getConversationById: (conversationId) => {
        const id = getConversationId(conversationId);
        return api.get(`/message-conversations/${id}`);
    },

    // Create a new conversation channel
    createConversation: (payload) => {
        const jsonPayload = buildConversationPayload(payload);
        return api.post("/message-conversations", jsonPayload);
    },

    // Update conversation metadata
    updateConversation: (conversationId, payload) => {
        const id = getConversationId(conversationId);
        const jsonPayload = buildConversationPayload(payload);
        return api.put(`/message-conversations/${id}`, jsonPayload);
    },

    // Delete conversation channel completely
    deleteConversation: (conversationId) => {
        const id = getConversationId(conversationId);
        return api.delete(`/message-conversations/${id}`);
    },

    // Mark all messages within a specific conversation as Read
    markConversationRead: (conversationId) => {
        const id = getConversationId(conversationId);
        return api.patch(`/message-conversations/${id}/read`);
    },

    // Dynamic Online/Offline status switch toggle
    updateOnlineStatus: (conversationId, isOnline) => {
        const id = getConversationId(conversationId);
        return api.patch(`/message-conversations/${id}/online-status`, { is_online: isOnline });
    },

    // Restore a soft-deleted conversation
    restoreConversation: (conversationId) => {
        const id = getConversationId(conversationId);
        return api.patch(`/message-conversations/restore/${id}`);
    },

    /*
    |--------------------------------------------------------------------------
    | MESSAGES CRUD (Actual Chat Bubbles)
    |--------------------------------------------------------------------------
    */

    // Fetch all existing system messages
    getAllMessages: () => api.get("/messages"),

    // Send/Post a brand new message inside a conversation
    createMessage: (payload) => {
        const jsonPayload = buildMessagePayload(payload);
        return api.post("/messages", jsonPayload);
    },

    // Load full chat history bubbles dynamically for the selected active user
    getMessagesByConversation: (conversationId) => {
        const id = getConversationId(conversationId);
        return api.get(`/messages/conversation/${id}`);
    },

    // Delete full chat history bubble logs from a conversation container
    deleteMessagesByConversation: (conversationId) => {
        const id = getConversationId(conversationId);
        return api.delete(`/messages/conversation/${id}`);
    },

    // Soft-delete/Undo specific single message bubble
    restoreMessage: (messageId) => {
        const id = getMessageId(messageId);
        return api.patch(`/messages/restore/${id}`);
    },

    // Get detail specs of a single message ID instance
    getMessageById: (messageId) => {
        const id = getMessageId(messageId);
        return api.get(`/messages/${id}`);
    },

    // Edit/Update text context inside a single message string bubble
    updateMessage: (messageId, payload) => {
        const id = getMessageId(messageId);
        const jsonPayload = buildMessagePayload(payload);
        return api.put(`/messages/${id}`, jsonPayload);
    },

    // Hard delete a single message record node item
    deleteMessage: (messageId) => {
        const id = getMessageId(messageId);
        return api.delete(`/messages/${id}`);
    },

    // Mark one single message specific state context as read
    markMessageRead: (messageId) => {
        const id = getMessageId(messageId);
        return api.patch(`/messages/${id}/read`);
    }
};

File: api/int-reztro-restaurant-mgt-dashboard-api-dev/app/services/message_service.py
from fastapi import HTTPException
from sqlalchemy import text
from sqlalchemy.exc import SQLAlchemyError


def safe_fetch_one(result):
    try:
        return result.mappings().first()
    except Exception:
        return None


# 🔹 CREATE
def create_message_service(db, payload):
    try:
        data = payload.model_dump()

        result = db.execute(
            text("""
                EXEC Reztro.AddMessage
                    @conversation_id=:conversation_id,
                    @sender_id=:sender_id,
                    @sender_type=:sender_type,
                    @message_type=:message_type,
                    @message_text=:message_text,
                    @attachment_url=:attachment_url,
                    @attachment_type=:attachment_type
            """),
            data
        )

        response = safe_fetch_one(result)
        db.commit()

        if response and response.get("success") == 0:
            raise HTTPException(status_code=400, detail=response.get("message"))

        return response

    except SQLAlchemyError as e:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 GET ALL
def get_messages_service(db):
    try:
        result = db.execute(text("EXEC Reztro.GetMessages"))
        return result.mappings().all()

    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 GET BY ID
def get_message_by_id_service(db, message_id: int):
    try:
        result = db.execute(
            text("""
                EXEC Reztro.GetMessageById
                    @message_id=:message_id
            """),
            {"message_id": message_id}
        )

        data = safe_fetch_one(result)

        if not data:
            raise HTTPException(status_code=404, detail="Message not found")

        return data

    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 GET BY CONVERSATION
def get_messages_by_conversation_service(db, conversation_id: int):
    try:
        result = db.execute(
            text("""
                EXEC Reztro.GetMessagesByConversationId
                    @conversation_id=:conversation_id
            """),
            {"conversation_id": conversation_id}
        )

        return result.mappings().all()

    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 UPDATE
def update_message_service(db, message_id: int, payload):
    try:
        params = payload.model_dump()
        params["message_id"] = message_id

        result = db.execute(
            text("""
                EXEC Reztro.UpdateMessage
                    @message_id=:message_id,
                    @message_text=:message_text,
                    @message_type=:message_type,
                    @attachment_url=:attachment_url,
                    @attachment_type=:attachment_type
            """),
            params
        )

        response = safe_fetch_one(result)
        db.commit()

        if response and response.get("success") == 0:
            raise HTTPException(status_code=400, detail=response.get("message"))

        return response

    except SQLAlchemyError as e:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 MARK MESSAGE AS READ
def mark_message_read_service(db, message_id: int):
    try:
        result = db.execute(
            text("""
                EXEC Reztro.MarkMessageAsRead
                    @message_id=:message_id
            """),
            {"message_id": message_id}
        )

        data = safe_fetch_one(result)
        db.commit()

        if not data:
            raise HTTPException(status_code=404, detail="Message not found")

        return data

    except SQLAlchemyError as e:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 MARK CONVERSATION AS READ
def mark_conversation_read_service(db, conversation_id: int):
    try:
        result = db.execute(
            text("""
                EXEC Reztro.MarkConversationMessagesAsRead
                    @conversation_id=:conversation_id
            """),
            {"conversation_id": conversation_id}
        )

        data = safe_fetch_one(result)
        db.commit()

        if not data:
            raise HTTPException(status_code=404, detail="Conversation not found")

        return data

    except SQLAlchemyError as e:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 DELETE SINGLE (SOFT)
def delete_message_service(db, message_id: int):
    try:
        result = db.execute(
            text("""
                EXEC Reztro.DeleteMessage
                    @message_id=:message_id
            """),
            {"message_id": message_id}
        )

        data = safe_fetch_one(result)
        db.commit()

        if not data:
            raise HTTPException(status_code=404, detail="Message not found")

        return data

    except SQLAlchemyError as e:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 DELETE BY CONVERSATION
def delete_messages_by_conversation_service(db, conversation_id: int):
    try:
        result = db.execute(
            text("""
                EXEC Reztro.DeleteMessagesByConversationId
                    @conversation_id=:conversation_id
            """),
            {"conversation_id": conversation_id}
        )

        data = safe_fetch_one(result)
        db.commit()

        if not data:
            raise HTTPException(status_code=404, detail="No messages found")

        return data

    except SQLAlchemyError as e:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )


# 🔹 RESTORE
def restore_message_service(db, message_id: int):
    try:
        result = db.execute(
            text("""
                EXEC Reztro.RestoreMessage
                    @message_id=:message_id
            """),
            {"message_id": message_id}
        )

        data = safe_fetch_one(result)
        db.commit()

        if not data:
            raise HTTPException(status_code=404, detail="Message not found")

        return data

    except SQLAlchemyError as e:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail=str(getattr(e, "orig", e))
        )